roles/kubernetes/preinstall/handlers/main.yml

```diff
@@ -32,4 +36,35 @@
   service:
     name: kubelet
     state: restarted
+  notify:
+    - Preinstall | kube-controller configured
+    - Preinstall | remove controller
+    - Preinstall | pause for controller
+    - Preinstall | restart controller
   when: "{{ inventory_hostname in groups['kube-master'] and not dns_early|bool }}"
+
+#
+# The following is used to restart the kube-controller to pick up
+# DNS changes in host_resolvconf mode.  This should be done much "better"
+# This is what seems to work now though.  This is only used when
+# kube-controller manifest is present and we are using host_resolvconf.
+#
+# This is triggered by a kubelet restart or network restart.
+#
+- name: Preinstall | kube-controller configured
+  stat: path="{{ kube_manifest_dir }}/kube-controller-manager.manifest"
+  register: kube_controller_set
+  when: "{{ inventory_hostname in groups['kube-master'] and dns_mode != 'none' and resolvconf_mode == 'host_resolvconf' }}"
+
+- name: Preinstall | remove controller
+  command: removes="{{ kube_manifest_dir }}/kube-controller-manager.manifest" creates=/tmp/jj mv "{{ kube_manifest_dir }}/kube-controller-manager.manifest" /tmp/jj
+  when: "{{ inventory_hostname in groups['kube-master'] and dns_mode != 'none' and resolvconf_mode == 'host_resolvconf' and kube_controller_set.stat.exists }}"
+
+- name: Preinstall | pause for controller
+  pause: seconds=50
+  when: "{{ inventory_hostname in groups['kube-master'] and dns_mode != 'none' and resolvconf_mode == 'host_resolvconf' and kube_controller_set.stat.exists }}"
+
+- name: Preinstall | restart controller
+  command: creates="{{ kube_manifest_dir }}/kube-controller-manager.manifest" removes=/tmp/jj mv /tmp/jj "{{ kube_manifest_dir }}/kube-controller-manager.manifest"
+  when: "{{ inventory_hostname in groups['kube-master'] and dns_mode != 'none' and resolvconf_mode == 'host_resolvconf' and kube_controller_set.stat.exists }}"
```

Comment:
If this is triggered BEFORE we define kube-controller-manager, you don't need this task. It will get re-established as it was before. however, it's a bit more obscure that way